prov/mrail/src/mrail_rma.c

```diff
@@ -0,0 +1,419 @@
+ * SOFTWARE.
+ */
+
+#include "mrail.h"
+
+static void mrail_subreq_to_rail(struct mrail_subreq *subreq, uint32_t rail,
+		struct iovec *out_iovs, void **out_descs,
+		struct fi_rma_iov *out_rma_iovs)
+{
+	const struct mrail_mr *mrail_mr;
+	struct mrail_addr_key *mr_map;
+	size_t i;
+
+	for (i = 0; i < subreq->iov_count; ++i) {
+		mrail_mr = subreq->descs[i];
+		//TODO: add base address from mrail_mr
+		out_iovs[i].iov_len	= subreq->iov[i].iov_len;
+		out_iovs[i].iov_base 	=
+			(void*)((uintptr_t)subreq->iov[i].iov_base);
+
+		out_descs[i] = (mrail_mr ?
+				fi_mr_desc(mrail_mr->rails[rail].mr) : NULL);
+	}
+
+	for (i = 0; i < subreq->rma_iov_count; ++i) {
+		mr_map = (struct mrail_addr_key *)subreq->rma_iov[i].key;
+		//TODO: add base address from mrail_addr_key
+		out_rma_iovs[i].addr 	= subreq->rma_iov[i].addr;
+		out_rma_iovs[i].len	= subreq->rma_iov[i].len;
+		out_rma_iovs[i].key	= mr_map[rail].key;
+	}
+}
+
+static ssize_t mrail_post_subreq(uint32_t rail,
+		struct mrail_subreq *subreq)
+{
+	ssize_t ret;
+	struct iovec rail_iov[MRAIL_IOV_LIMIT];
+	void *rail_descs[MRAIL_IOV_LIMIT];
+	struct fi_rma_iov rail_rma_iov[MRAIL_IOV_LIMIT];
+	struct fi_msg_rma msg;
+
+	struct mrail_req *req = subreq->parent;
+	struct mrail_ep *mrail_ep = req->mrail_ep;
+
+	uint64_t flags = req->flags;
+
+	mrail_subreq_to_rail(subreq, rail, rail_iov, rail_descs, rail_rma_iov);
+
+	msg.msg_iov		= rail_iov;
+	msg.desc		= rail_descs;
+	msg.iov_count		= subreq->iov_count;
+	msg.addr		= req->remote_addrs[rail];
+	msg.rma_iov		= rail_rma_iov;
+	msg.rma_iov_count	= subreq->rma_iov_count;
+	msg.context		= &subreq->context;
+
+	if (req->op_type == FI_READ) {
+		ret = fi_readmsg(mrail_ep->rails[rail].ep, &msg, flags);
+	} else {
+		/* Immediate data is sent with the last subreq only */
+		if (flags & FI_REMOTE_CQ_DATA) {
+			if (req->pending_subreq > 0) {
+				flags &= ~FI_REMOTE_CQ_DATA;
+			} else {
+				msg.data = req->data;
+			}
+		}
+		ret = fi_writemsg(mrail_ep->rails[rail].ep, &msg, flags);
+	}
+
+	return ret;
+}
+
+static ssize_t mrail_post_req(struct mrail_req *req)
+{
+	size_t i;
+	uint32_t rail;
+	ssize_t ret = 0;
+
+	while (req->pending_subreq >= 0) {
+		/* Try all rails before giving up */
+		for (i = 0; i < req->mrail_ep->num_eps; ++i) {
+			rail = mrail_get_tx_rail(req->mrail_ep);
+
+			ret = mrail_post_subreq(rail,
+					&req->subreqs[req->pending_subreq]);
+			if (ret != -FI_EAGAIN) {
+				break;
+			} else {
+				/* One of the rails is busy. Try progressing. */
+				mrail_poll_cq(req->mrail_ep->util_ep.tx_cq);
+			}
+		}
+
+		if (ret != 0) {
+			if (ret == -FI_EAGAIN) {
+				break;
+			}
+			/* TODO: Handle errors besides FI_EAGAIN */
+			assert(0);
+		}
+		req->pending_subreq--;
+	}
+
+	return ret;
+}
+
+static struct mrail_req *mrail_dequeue_deferred_req(struct mrail_ep *mrail_ep)
+{
+	struct slist_entry *e;
+
+	e = slist_remove_head(&mrail_ep->deferred_reqs);
+	if (!e) {
+		return NULL;
+	}
+	return container_of(e, struct mrail_req, entry);
+}
```

Comment:
This function is barely needed.  We can add '_container' functions to slist, similar to what is available for dlist.